storage/remote/write_handler.go

```diff
@@ -256,87 +269,178 @@ func (h *writeHandler) write(ctx context.Context, req *prompb.WriteRequest) (err
+				errors.Is(err, storage.ErrDuplicateSampleForTimestamp) ||
+				errors.Is(err, storage.ErrTooOldSample) {
+				// TODO(bwplotka): Not too spammy log?
+				level.Error(h.logger).Log("msg", "Out of order sample from remote write", "err", err.Error(), "series", ls.String(), "timestamp", s.Timestamp)
+				badRequestErrs = append(badRequestErrs, fmt.Errorf("%w for series %v", err, ls.String()))
+				continue
+			}
+			return err, http.StatusInternalServerError
 		}
 
+		// Native Histograms.
+		for _, hp := range ts.Histograms {
+			if hp.IsFloatHistogram() {
+				ref, err = app.AppendHistogram(ref, ls, hp.Timestamp, nil, hp.ToFloatHistogram())
+			} else {
+				ref, err = app.AppendHistogram(ref, ls, hp.Timestamp, hp.ToIntHistogram(), nil)
+			}
+			if err == nil {
+				rs.histograms++
+				continue
+			}
+			// Handle append error.
+			// Although AppendHistogram does not currently return ErrDuplicateSampleForTimestamp there is
+			// a note indicating its inclusion in the future.
+			if errors.Is(err, storage.ErrOutOfOrderSample) ||
+				errors.Is(err, storage.ErrOutOfBounds) ||
+				errors.Is(err, storage.ErrDuplicateSampleForTimestamp) {
+				// TODO(bwplotka): Not too spammy log?
+				level.Error(h.logger).Log("msg", "Out of order histogram from remote write", "err", err.Error(), "series", ls.String(), "timestamp", hp.Timestamp)
+				badRequestErrs = append(badRequestErrs, fmt.Errorf("%w for series %v", err, ls.String()))
+				continue
+			}
+			return err, http.StatusInternalServerError
+		}
+
+		// Exemplars.
 		for _, ep := range ts.Exemplars {
 			e := ep.ToExemplar(&b, req.Symbols)
-			h.appendExemplar(timeLimitApp, e, ls, &outOfOrderExemplarErrs)
-		}
+			ref, err = app.AppendExemplar(ref, ls, e)
+			if err == nil {
+				rs.exemplars++
+				continue
 
-		err = h.appendHistogramsV2(timeLimitApp, ts.Histograms, ls)
-		if err != nil {
-			return err
+			}
+			// Handle append error.
+			// TODO(bwplotka): I left the logic as in v1, but we might want to make it consistent with samples and histograms.
+			// Since exemplar storage is still experimental, we don't fail in anyway, the request on ingestion errors.
+			if errors.Is(err, storage.ErrOutOfOrderExemplar) {
+				outOfOrderExemplarErrs++
+				level.Debug(h.logger).Log("msg", "Out of order exemplar", "err", err.Error(), "series", ls.String(), "exemplar", fmt.Sprintf("%+v", e))
+				continue
+			}
+			level.Debug(h.logger).Log("msg", "Error while adding exemplar in AppendExemplar", "series", ls.String(), "exemplar", fmt.Sprintf("%+v", e), "err", err)
 		}
 
 		m := ts.ToMetadata(req.Symbols)
-		if _, err = timeLimitApp.UpdateMetadata(0, ls, m); err != nil {
+		if ref, err = app.UpdateMetadata(ref, ls, m); err != nil {
 			level.Debug(h.logger).Log("msg", "error while updating metadata from remote write", "err", err)
+			// Metadata is attached to each series, so since Prometheus does not reject sample without metadata information,
+			// we don't report remote write error either.
```

Comment:
should we have a metric for write handler samples received _without_ metadata, I know some people have been confused with the current v1 implementation that they don't see metadata for the samples they've written to prometheus via the write handler

Reply:
No harm in adding sure, but I think all the logs here (potentially spammy ones) should be metrics, so the whole handler instrumentation needs redesign--too big for this PR.